### ISSUE TITLE
'Stream read error' thrown when attempting to read any mgx2 file

### ISSUE THREAD

**Issue by boombatower** [goto-bus-stop/recanalyst#4]
```
::1:52854 [500]: /test.php - Uncaught exception 'Exception' with message 'Stream read error' in /home/boombatower/software/recanalyst/src/Stream.php:125
Stack trace:
#0 /home/boombatower/software/recanalyst/src/MemoryStream.php(167): RecAnalyst\Stream->readBuffer('\x00\x00\x00 \x01\x00\x00\x00\x02\x00\x00\x00\v\x00\x00...', 16777216)
#1 /home/boombatower/software/recanalyst/src/RecAnalyst.php(533): RecAnalyst\MemoryStream->readString('\x00\x00\x00 \x01\x00\x00\x00\x02\x00\x00\x00\v\x00\x00...')
#2 /home/boombatower/software/recanalyst/src/RecAnalyst.php(1354): RecAnalyst\RecAnalyst->analyzeHeaderStream()
#3 /home/boombatower/software/recanalyst/test.php(16): RecAnalyst\RecAnalyst->analyze()
#4 {main}
  thrown in /home/boombatower/software/recanalyst/src/Stream.php on line 125
```

Looks like it's getting the player name string length wrong and trying to read way too many bytes. Obviously, that could mean the parser got off track long before that.

**Comment by goto-bus-stop** [comment]
I'm guessing that it's only files from The Forgotten 4.0 onwards? I'll have a look tonight probably.

**Comment by jduyon** [comment]
Yeah, I think they changed the header format in v4, this doesn't work for the new .msx files.

**Comment by goto-bus-stop** [comment]
Sorry for taking so long. Doesn't look like I'll have time to look into this soon either…If anyone else is willing to attempt to add 4.0 support in in the mean time, would be very welcome :stuck_out_tongue_winking_eye:

**Comment by jduyon** [comment]
I'm new to this recanalyst repo, and its taking me a while to get familiar with the new file format. I'm actively working on it though!

**Comment by coffenbacher** [comment]
Did anyone have any luck with this???

**Comment by goto-bus-stop** [comment]
Working on this now. msx(2) files save a bunch more properties, so there's a few more bytes that RecAnalyst has to skip. (for now. ideally it'd just read them properly too. :smile: )

I don't have many recorded games to test with atm, if anyone has any `.msx(2)`s to spare that'd be really helpful :)

**Comment by goto-bus-stop** [comment]
https://github.com/goto-bus-stop/recanalyst/tree/msx

^ works for at least those few .msx2 files that I grabbed from AoCzone.

I'll try some `.mgx2`s as well, because the error message in @boombatower's post seems to occur at a different place than the ones I got with `.msx2`…

**Comment by jduyon** [comment]
I was able to parse an MSX multiplayer file (I'm attaching it here).


[mp_4players_2teams_2AI_20150911.zip](https://github.com/goto-bus-stop/recanalyst/files/72230/mp_4players_2teams_2AI_20150911.zip)

Not sure if you've noticed, but it looks like the new recorded game file for the expansion for AOE2HD, has a new extension now (i think it was .AGE or something like that).

**Comment by goto-bus-stop** [comment]
Thanks!

`mgx2` files are fixed by much the same logic as `msx(2)`. Just need to figure out the mapping between the version-ish number in the recorded game files and the HD patch versions, and then I think mgx2/msx/msx2 support is good to go.

**Comment by goto-bus-stop** [comment]
Very sorry for the months-long delay. `master` now supports (most?) mgx2 and msx(2) files. The "new" aoe2record files are still somewhat of an issue, but that's moving to #9 :)

I don't really have a release process in place for RecAnalyst, I promised myself in 2014 that I'd publish to composer once I finished the internals refactor that's going on in https://github.com/goto-bus-stop/recanalyst/tree/api, but lol.
